File: backend/app/models/schemas.py
from typing import Literal

from pydantic import BaseModel, Field


class ChatMessage(BaseModel):
    role: Literal["user", "assistant", "system"]
    content: str


class ProductQuery(BaseModel):
    query: str = Field(..., min_length=1)
    category: str | None = None
    min_price: float | None = None
    max_price: float | None = None
    brand: str | None = None
    sources: list[str] | None = None


BuySuggestion = Literal["frequent", "restock", "recent", "new"]
Origin = Literal["catalog", "live"]


class ProductListing(BaseModel):
    id: str
    title: str
    source: str
    # Where this listing came from: "catalog" (Salesforce) or "live" (fetched
    # directly from the store's website). Live results are highlighted as new.
    origin: Origin | None = None
    current_price: float | None = None
    original_price: float | None = None
    last_purchased_price: float | None = None
    discount: int | None = None
    rating: str | None = None
    review_count: int | None = None
    rank: int | None = None
    product_url: str | None = None
    image_url: str | None = None
    availability: str | None = None
    weight: str | None = None
    last_ordered_date: str | None = None
    times_purchased: int | None = None
    buy_suggestion: BuySuggestion | None = None
    suggestion_reason: str | None = None


class ChatRequest(BaseModel):
    messages: list[ChatMessage] = Field(..., min_length=1)


class ChatResponse(BaseModel):
    reply: str
    product_query: ProductQuery | None = None


class ProductSearchResponse(BaseModel):
    results: list[ProductListing]


class RecommendationRequest(BaseModel):
    # Optional free-text preference; blank is coerced to the default by the router.
    user_input: str = "Give recommendations"
    # When true, bypass the server-side cache and re-fetch from the engine.
    refresh: bool = False


class RecommendationItem(BaseModel):
    product_name: str
    product_url: str | None = None
    price: float | None = None
    reasoning: str | None = None
    rating: str | None = None
    highlights: list[str] = []


class RecommendationResponse(BaseModel):
    insight_message: str
    recommendations: list[RecommendationItem]


class CartCheckoutRequest(BaseModel):
    products: list[str] = Field(..., min_length=1)


class CartCheckoutResponse(BaseModel):
    submitted: int  # count of products sent
    detail: str  # human-readable confirmation


class AgentCartItem(BaseModel):
    id: str  # f"{source}:{title}" — cart membership key, matches the FE CartItem shape
    name: str
    source: str | None = None


class AgentResponse(BaseModel):
    reply: str
    results: list[ProductListing] = []
    cart: list[AgentCartItem] = []
    checkout: CartCheckoutResponse | None = None


class RefreshRequest(BaseModel):
    source: Literal["amazon", "flipkart"]


class RefreshResponse(BaseModel):
    detail: str  # human-readable confirmation


class OtpRequest(BaseModel):
    otp: str = Field(..., min_length=1)


class OtpResponse(BaseModel):
    detail: str  # human-readable confirmation

File: worker/src/models/schemas.ts
// Wire types — snake_case field names to keep the exact FE↔BE JSON contract
// (mirrors backend/app/models/schemas.py and frontend/src/lib/types.ts).

export type Role = "user" | "assistant" | "system";
export type BuySuggestion = "frequent" | "restock" | "recent" | "new";
export type Origin = "catalog" | "live";

export interface ChatMessage {
  role: Role;
  content: string;
}

export interface ProductQuery {
  query: string;
  category?: string | null;
  min_price?: number | null;
  max_price?: number | null;
  brand?: string | null;
  sources?: string[] | null;
}

export interface ProductListing {
  id: string;
  title: string;
  source: string;
  origin?: Origin | null;
  current_price?: number | null;
  original_price?: number | null;
  last_purchased_price?: number | null;
  discount?: number | null;
  rating?: string | null;
  review_count?: number | null;
  rank?: number | null;
  product_url?: string | null;
  image_url?: string | null;
  availability?: string | null;
  weight?: string | null;
  last_ordered_date?: string | null;
  times_purchased?: number | null;
  buy_suggestion?: BuySuggestion | null;
  suggestion_reason?: string | null;
}

export interface ChatRequest {
  messages: ChatMessage[];
  thread_id?: string | null;
}

export interface ProductSearchResponse {
  results: ProductListing[];
}

export interface RecommendationRequest {
  user_input?: string;
  refresh?: boolean;
}

export interface RecommendationItem {
  product_name: string;
  product_url?: string | null;
  price?: number | null;
  reasoning?: string | null;
  rating?: string | null;
  highlights?: string[];
  image_url?: string | null;
}

export interface RecommendationResponse {
  insight_message: string;
  recommendations: RecommendationItem[];
}

export interface CartItemCheckout {
  name: string;
  source?: string | null;
}

export interface CartCheckoutRequest {
  products: Array<string | CartItemCheckout>;
}

export interface CartCheckoutResponse {
  submitted: number;
  detail: string;
}

export interface AgentCartItem {
  id: string;
  name: string;
  source?: string | null;
}

export interface PendingLive {
  query: string;
  sources: string[];
  min_price?: number | null;
  max_price?: number | null;
}

export interface AgentResponse {
  reply: string;
  results: ProductListing[];
  cart: AgentCartItem[];
  checkout?: CartCheckoutResponse | null;
  pending_live?: PendingLive | null;
  thread_id?: string | null;
}

export interface RefreshRequest {
  source: "amazon" | "flipkart";
}

export interface RefreshResponse {
  detail: string;
}

export interface OtpResponse {
  detail: string;
}

export interface IdentifyRequest {
  image: string; // base64 (no data-URL prefix)
  mime_type?: string;
}

export interface MustHaveProduct {
  id: string;
  title: string;
  source: string;
  reason?: string | null;
}

export interface IdentifyResponse {
  reply: string;
  results: ProductListing[];
  must_have: MustHaveProduct[];
}
